File: python/day24.py
from collections import defaultdict, deque

with open('./inputs/24.txt') as f:
    instructions = f.readlines()

connections = defaultdict(set)

for line in instructions:
    a, b = map(int, line.split('/'))
    connections[a].add((a, b))
    connections[b].add((a, b))

start = ({(0, 0)}, 0, 0)
que = deque([start])

max_score = 0
longest_strongest = (0, 0)

while que:
    path, score, conn = que.popleft()
    if score > max_score:
        max_score = score
    if (len(path), score) > longest_strongest:
        longest_strongest = (len(path), score)
    for candidate in connections[conn]:
        if candidate not in path:
            new_connection = candidate[candidate.index(conn) - 1]
            new_score = score + sum(candidate)
            new_path = path | {candidate}
            que.append((new_path, new_score, new_connection))

print(max_score)
print(longest_strongest[-1])

File: nim/day24.nim
import strutils, sequtils, math

const instructions = readFile("./inputs/24.txt").splitLines
var
  connections = newSeq[array[2, int]]()
  maxScore: int
  longestStrongest = (0, 0)
  doubles = newSeq[int]()

for i, line in instructions:
  let pins = line.split('/').map(parseInt)
  if pins[0] != pins[1]: connections.add([pins[0], pins[1]])
  else: doubles.add(pins[0])

var visited: set[int8] = {}


proc calcScore(p: seq[int]) =
  if p.len == 0: return
  var
    used: set[int8] = {}
    score: int
    length = p.len
  for n in p:
    score += 2*n
    if (n in doubles) and (n.int8 notin used):
      score += 2*n
      inc length
      used.incl(n.int8)
  score -= p[^1]
  maxScore = max(maxScore, score)
  longestStrongest = max(longestStrongest, (length, score))


iterator allMatches(currentNode: int): tuple[conn, pos: int8] =
  for i, bridge in connections:
    let f = bridge.find(currentNode).int8
    if (f > -1) and (i.int8 notin visited):
      yield (i.int8, 1-f)


proc solve(path: var seq[int], currentNode = 0) =
  for match in currentNode.allMatches:
    let newConnection = connections[match.conn][match.pos]
    visited.incl match.conn

    path.add(newConnection)
    path.solve(newConnection)
    discard path.pop()

    visited.excl match.conn
  calcScore(path)


var path: seq[int] = @[]
path.solve()

echo maxScore
echo longestStrongest[1]
